presto-accumulo/src/main/java/com/facebook/presto/accumulo/io/AccumuloRecordCursor.java

```
@@ -0,0 +1,584 @@
+import static com.facebook.presto.spi.type.TimeType.TIME;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implementation of Presto RecordCursor, responsible for iterating over a Presto split, reading
+ * rows of data and then implementing various methods to retrieve columns within each row.
+ *
+ * @see AccumuloRecordSet
+ * @see AccumuloRecordSetProvider
+ */
+public class AccumuloRecordCursor
+        implements RecordCursor
+{
+    private final List<AccumuloColumnHandle> cHandles;
+    private final String[] fieldToColumnName;
+
+    private long bytesRead = 0L;
+    private long nanoStart = 0L;
+    private long nanoEnd = 0L;
+    private final BatchScanner scan;
+    private final Iterator<Entry<Key, Value>> iterator;
+    private final AccumuloRowSerializer serializer;
+    private Entry<Key, Value> prevKV;
+    private Text prevRowID = new Text();
+    private Text rowID = new Text();
+
+    /**
+     * Creates a new instance of {@link AccumuloRecordCursor}
+     *
+     * @param serializer Serializer to decode the data stored in Accumulo
+     * @param scan BatchScanner for retrieving rows of data
+     * @param rowIdName Presto column that is the Accumulo row ID
+     * @param cHandles List of column handles in each row
+     * @param constraints List of all column constraints
+     */
+    public AccumuloRecordCursor(AccumuloRowSerializer serializer, BatchScanner scan,
+            String rowIdName, List<AccumuloColumnHandle> cHandles,
+            List<AccumuloColumnConstraint> constraints)
+    {
+        this.cHandles = requireNonNull(cHandles, "cHandles is null");
+        this.scan = requireNonNull(scan, "scan is null");
+
+        this.serializer = requireNonNull(serializer, "serializer is null");
+        this.serializer.setRowIdName(rowIdName);
+
```

You probably want to check `requireNonNull` for constraints here to fail early.